cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+      boolean send = false;
+      cancelPollingAndScheduleNext();
+      synchronized (lastNotificationLock) {
+        if (lastNotification == null || notification.getTimestamp() > lastNotification.getTimestamp()) {
+          send = true;
+          lastNotification = notification;
+        }
+      }
+      if (send) {
+        sendNotificationToActiveTasks(notification);
+      }
+    }
+
+    /**
+     * Send a {@link StreamSizeNotification} to all the active {@link StreamSizeSchedule} referenced
+     * by this object.
+     */
+    private void sendNotificationToActiveTasks(final StreamSizeNotification notification) {
+      List<StreamSizeScheduleTask> values;
+      synchronized (this) {
+        values = ImmutableList.copyOf(scheduleTasks.values());
+      }
+
+      for (final StreamSizeScheduleTask task : values) {
+        if (!task.isRunning()) {
+          continue;
+        }
+        notificationExecutor.execute(new Runnable() {
+          @Override
+          public void run() {
+            task.received(notification);
+          }
+        });
+      }
+    }
+
+    private Id.NotificationFeed getFeed() {
+      return new Id.NotificationFeed.Builder()
+        .setNamespaceId(streamId.getNamespaceId())
+        .setCategory(Constants.Notification.Stream.STREAM_FEED_CATEGORY)
+        .setName(String.format("%sSize", streamId.getName()))
+        .build();
+    }
+
+    /**
+     * Cancel the currently scheduled stream size polling task, and reschedule one for later.
+     */
+    private void cancelPollingAndScheduleNext() {
+      // This method might be called from the call to #received defined in the below Runnable - in which case
+      // this scheduledPolling would in fact be running. Hence we don't want to interrupt the running task
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(false);
+      }
+
+      // Regardless of whether cancelling was successful, we still want to schedule the next polling
+      scheduledPolling = streamPollingExecutor.schedule(createPollingRunnable(), pollingDelay, TimeUnit.SECONDS);
+    }
+
+    /**
+     * @return a runnable that uses the {@link StreamAdmin} to poll the stream size, and creates a fake notification
+     *         with that size, so that this information can be treated as if it came from a real notification.
+     */
+    private Runnable createPollingRunnable() {
+      return new Runnable() {
+        @Override
+        public void run() {
+          // We only perform polling if at least one scheduleTask is active
+          if (activeTasks.get() > 0) {
+            long size = pollStream();
+
+            // We don't need a notification context here
+            received(new StreamSizeNotification(System.currentTimeMillis(), size), null);
+          }
+        }
+      };
+    }
+
+    /**
+     * @return size of the stream queried directly from the file system
+     */
+    private long pollStream() {
```

Comment:
This will come after this PR https://github.com/caskdata/cdap/pull/1364 is merged.